### ISSUE TITLE
commonStrict mode does not import module object properly

### ISSUE THREAD

**Issue by ikarienator** [6to5/6to5#675]
IMO, there is no reason for `commonStrict` mode to check the `__esModule` hack.

**Comment by sebmck** [comment]
It'll still be necessary for compatibility with Traceur/es6-module-transpiler generated modules.

**Comment by sebmck** [comment]
If you're concerned about the size of the helpers then the optional [selfContained](6to5.org/docs/usage/transformers/#self-contained) transformer may be of interest.

**Comment by ikarienator** [comment]
I'm not concerning about the size of the helpers. I have some es6 modules compliant to the ES6 spec but without a `__esModule` flag exported. The generated code of 6to5:

```js
var _interopRequireWildcard = function (obj) { return obj && obj.__esModule ? obj : { "default": obj }; };
```

will not properly import the module.

And why is that necessary to make it compatible with the broken behavior of Traceur or es6-module-transpiler? (For the record, I think you are doing much better than them!) If it is Traceur who's generating the wrong code, then it is Traceur's responsibility to fix their own code. I should be able to assume that `commonStrict` means to be strictly compliant to ES6 spec instead of being compatible with some broken Google implementation of ES6 transpiler or some random library on the interweb.

To be fair, compatibility with other libraries is important. But we should not sacrifice correctness for it given that it is already been handled by the default CommonJS mode.

**Comment by sebmck** [comment]
@ikarienator I guess I could add this behaviour to the `commonStrict` module formatter but it's going to break importing the defaults of all CommonJS modules.

**Comment by sebmck** [comment]
```javascript
import express from "express";
```

is going to have to be transpiled to

```javascript
var express = require("express")["default"];
```

**Comment by ikarienator** [comment]
That's the point. `commonStrict` is supposed to strictly follow the spec, no? I am not concerned about compatibility with existing CommonJS modules. If I was, I would use the `common` module formatter. cc @michaelficarra

**Comment by sebmck** [comment]
@ikarienator It doesn't make sense to convert ES6 modules to CommonJS without any interop. All of the strict transformers just opt out of the `module.exports` behaviour.

**Comment by ikarienator** [comment]
Interop with CommonJS is a plus. It cannot come with the cost of preventing me from importing a legitimate ES6 module.

**Comment by sebmck** [comment]
@ikarienator Why do you keep on referencing "legitimate" ES6 modules? There are none in the wild and if you're so concerned about spec compliancy then you wouldn't be using an intermediary layer as CommonJS is fundamentally different to ES6 modules.

**Comment by michaelficarra** [comment]
@sebmck:

> There are none in the wild

That is incorrect. Spec-compliant modules exist in the wild because they were compiled using 6to5 1.x. We cannot interoperate with these modules, even though they are following the standard. We could re-compile them with 6to5 3.x, but that would cause a breaking interface change. And what would we get from it? Spec non-conformance.

We are going to write our own custom module formatter for strict ES6 compliance. When it is complete, we'll send a PR.

**Comment by thejameskyle** [comment]
@michaelficarra What was in 1.x was far from spec-compliant. Spec-compliancy and interop is impossible in common.js

**Comment by sebmck** [comment]
Sure but when you can't support live bindings and circular dependencies you
have to do the best you can to maintain that interop.

I'm not against making this the behaviour of the strict module formatters
but it'd be a breaking change and we only just bumped majors so the next
one is a little way off.

More than open to a PR just note though if you're basing it off the
internal module formatters that they're very interconnected and could use
some refactoring so apologies for the shoddy code.

On Wednesday, 4 February 2015, Michael Ficarra <notifications@github.com>
wrote:

> @sebmck <https://github.com/sebmck>:
>
> There are none in the wild
>
> That is incorrect. Spec-compliant modules exist in the wild because they
> were compiled using 6to5 1.x. We cannot interoperate with these modules,
> even though they are following the standard. We could re-compile them with
> 6to5 3.x, but that would cause a breaking interface change. And what would
> we get from it? Spec non-conformance.
>
> We are going to write our own custom module formatter for strict ES6
> compliance. When it is complete, we'll send a PR.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/6to5/6to5/issues/675#issuecomment-72749385>.
>


-- 
Sebastian McKenzie

**Comment by sebmck** [comment]
Also apologies for the bad initial tone, definently didn't approach this
with the right attitude.

On Wednesday, 4 February 2015, Sebastian McKenzie <sebmck@gmail.com> wrote:

> Sure but when you can't support live bindings and circular dependencies
> you have to do the best you can to maintain that interop.
>
> I'm not against making this the behaviour of the strict module formatters
> but it'd be a breaking change and we only just bumped majors so the next
> one is a little way off.
>
> More than open to a PR just note though if you're basing it off the
> internal module formatters that they're very interconnected and could use
> some refactoring so apologies for the shoddy code.
>
> On Wednesday, 4 February 2015, Michael Ficarra <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> @sebmck <https://github.com/sebmck>:
>>
>> There are none in the wild
>>
>> That is incorrect. Spec-compliant modules exist in the wild because they
>> were compiled using 6to5 1.x. We cannot interoperate with these modules,
>> even though they are following the standard. We could re-compile them with
>> 6to5 3.x, but that would cause a breaking interface change. And what would
>> we get from it? Spec non-conformance.
>>
>> We are going to write our own custom module formatter for strict ES6
>> compliance. When it is complete, we'll send a PR.
>>
>> --
>> Reply to this email directly or view it on GitHub
>> <https://github.com/6to5/6to5/issues/675#issuecomment-72749385>.
>>
>
>
> --
> Sebastian McKenzie
>


-- 
Sebastian McKenzie

**Comment by michaelficarra** [comment]
@thejameskyle: We're not concerned about CommonJS interop. If we need to support that, we are certainly in a CommonJS environment, so we will just use `require` directly. For everything else, we can use ES6 imports/exports. Also, in what ways was it non-compliant?

@sebmck: No problem, we could also have been more clear about our goals.

**Comment by thejameskyle** [comment]
@michaelficarra mostly around live bindings and circular dependencies that @sebmck mentioned.

**Comment by sebmck** [comment]
@thejameskyle That's still not supported because it can't be in 2.x and 3x. (Live bindings **can** be implemented to some extent)

**Comment by thejameskyle** [comment]
@sebmck I'm just saying it's impossible to make it supported without rebuilding something like system.js in common.js.

**Comment by ikarienator** [comment]
I think I'm confused about the goal of the `commonStrict` module formatter. Can someone explain it?
If compliance is not a concern, what is the principle of constructing this formatter?

**Comment by sebmck** [comment]
@ikarienator It was originally to opt out of the module interop in 2.x. Since 3.x changes that interop to be more inline with the spec and expected behaviour the strict module formatters were basically neutered because most of what they did wasn't necessary anymore. The behaviour you propose should likely be the actual behaviour of the strict module formatters as it mirrors the behaviour of the current ones but this wasn't always the case.

**Comment by ikarienator** [comment]
But soon enough people will start to pick up the standard behavior and this will be become an issue. Do you have the figure on how many people are using the `commonStrict` mode? If not too many then I believe changing it to comply to the spec has long term benefits.

**Comment by sebmck** [comment]
@ikarienator No idea who uses what. I've removed features in the past that I didn't think anyone used but ended up getting a lot of requests to bring them back (optional runtime is one). I'm completely open to making this change but only in the next major.

**Comment by sebmck** [comment]
I can temporarily add another module formatter if you want until it becomes the default behaviour of the strict formatters, just need a name for it.

**Comment by ikarienator** [comment]
That will be awesome to have! For now I think we can rebuild all our libraries in 3x. That will fix the problem for a while. But I think @michaelficarra is not happy with that in the long run. What do you think , @michaelficarra?

**Comment by michaelficarra** [comment]
@sebmck: That would be great! How about `standard`?

**Comment by sebmck** [comment]
That should do the trick. I'll push out a new minor in a bit.

**Comment by sebmck** [comment]
`commonStandard` is available as of 3.4.0. In 4.x I'll merge it with the strict module formatters and alias `commonStandard` to `commonStrict`. Let me know if there's any issues with it as I'll happily break/change it in a patch since noone's relying on it yet. Thanks!
